### PR TITLE
Create our own comment parser for codegen

### PR DESCRIPTION
**Pull request by whaught**
Issue #1226

Per comments in PR #1226 there was some dissatisfaction with our comment parsing.
This parser allows for codegen tags of the form` +foo:key=value,key2=value2`

We currently cheat by using the k8s parser and search for the prefix `+foo:key` to achieve such a result, but that doesn't allow for multiple comma-separated values per line.

Although this is a break from the current parser: https://github.com/kubernetes/gengo/blob/master/types/comments.go
We're already kind of abusing it with our "+genreconciler:class" prefix search.

This also reduces the number of times we fully scan the comments.

### CODE REVIEW

File: codegen/cmd/injection-gen/generators/comment_parser.go [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+			continue
+		}
+
+		options := strings.SplitN(line[len(marker):], ":", 2)
+		if len(options) == 2 {
+			vals := strings.Split(options[1], ",")
+
+			opts := out[options[0]]
+			if opts == nil {
+				opts = map[string]string{}
```
Comment (vagababov) [inline codegen/cmd/injection-gen/generators/comment_parser.go:56]:
```suggestion
				opts = make(map[string]string, len(vals))
```

File: codegen/cmd/injection-gen/generators/comment_parser_test.go [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package generators
+
+import "testing"
+
+func TestParseComments(t *testing.T) {
+
```
Comment (vagababov) [inline codegen/cmd/injection-gen/generators/comment_parser_test.go:21]:
```suggestion
```

File: codegen/cmd/injection-gen/generators/comment_parser_test.go [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+		t.Errorf("Failed to extract single key got=%t,%v want=true,{\"key\":\"value\"}", ok, val)
+	}
+
+	if val, ok := extracted["manypairs"]; !ok || val["key1"] != "value1" || val["key2"] != "value2" {
+		t.Errorf("Failed to extract single key got=%t,%v want=true,{\"key\":\"value\"}", ok, val)
+	}
+}
+
+func TestMergeDuplicates(t *testing.T) {
+
```
Comment (vagababov) [inline codegen/cmd/injection-gen/generators/comment_parser_test.go:56]:
```suggestion
```

File: codegen/cmd/injection-gen/generators/packages.go [hunk 4]
```diff
@@ -188,22 +188,27 @@ func MustParseClientGenTags(lines []string) Tags {
...
-	comments := append(append([]string{}, t.SecondClosestCommentLines...), t.CommentLines...)
-	_, nonNamespaced := types.ExtractCommentTags("+", comments)["genclient:nonNamespaced"]
-	return nonNamespaced
+func isNonNamespaced(tags map[string]map[string]string) bool {
+	vals, has := tags["genclient"]
+	if !has {
+		return false
+	}
+	_, has = vals["nonNamespaced"]
+	return has
```
Comment (vagababov) [inline codegen/cmd/injection-gen/generators/packages.go:211]:
```suggestion
	return has && vals["nonNamespaced"]
```

File: codegen/cmd/injection-gen/generators/comment_parser.go [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+			}
+
+			for _, pair := range vals {
+				kv := strings.SplitN(pair, "=", 2)
+				if len(kv) == 2 {
```
Comment (vagababov) [inline codegen/cmd/injection-gen/generators/comment_parser.go:61]:
```suggestion
				if kv := strings.SplitN(pair, "=", 2); len(kv) == 2 {
```

File: codegen/cmd/injection-gen/generators/comment_parser.go [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+					opts[kv[0]] = kv[1]
+				} else if len(kv) == 1 {
+					opts[kv[0]] = ""
+				}
+			}
+			out[options[0]] = opts
+
```
Comment (vagababov) [inline codegen/cmd/injection-gen/generators/comment_parser.go:68]:
```suggestion
```

File: codegen/cmd/injection-gen/generators/comment_parser.go [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+func ExtractCommentTags(marker string, lines []string) map[string]map[string]string {
+	out := map[string]map[string]string{}
+	for _, line := range lines {
+		line = strings.Trim(line, " ")
+		if len(line) == 0 {
+			continue
+		}
+		if !strings.HasPrefix(line, marker) {
+			continue
+		}
```
Comment (vagababov) [inline codegen/cmd/injection-gen/generators/comment_parser.go:48]:
```suggestion
		if len(line) == 0 || !strings.HasPrefix(line, marker){
			continue
		}
```

File: codegen/cmd/injection-gen/generators/comment_parser.go [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+//   +bar
+//
+// Then this function will return:
+//   map[string]map[string]string{"foo":{"key":value1","key2":"value2"}, "bar": {}}
+//
+// Users are not expected to repeat values.
+func ExtractCommentTags(marker string, lines []string) map[string]map[string]string {
+	out := map[string]map[string]string{}
+	for _, line := range lines {
+		line = strings.Trim(line, " ")
```
Comment (vagababov) [inline codegen/cmd/injection-gen/generators/comment_parser.go:42]: https://golang.org/pkg/strings/#TrimSpace, perhaps. What if it's a tab?

File: codegen/cmd/injection-gen/generators/packages.go (hunk 4 above)
Comment (vagababov) [reply codegen/cmd/injection-gen/generators/packages.go:211]: Looks the same.

**Comment by antoineco** [comment]
LGTM :+1:

File: codegen/cmd/injection-gen/generators/comment_parser.go [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+		if len(line) == 0 || !strings.HasPrefix(line, marker){
```
Comment (mattmoor) [inline codegen/cmd/injection-gen/generators/comment_parser.go:43]: Format Go code:
```suggestion
		if len(line) == 0 || !strings.HasPrefix(line, marker) {
```

File: codegen/cmd/injection-gen/generators/comment_parser.go (hunk 8 above)
Comment (whaught) [reply codegen/cmd/injection-gen/generators/comment_parser.go:42]: This was all just copying the existing k8s parser behavior. Sure let's trimspace

File: codegen/cmd/injection-gen/generators/packages.go (hunk 4 above)
Comment (whaught) [reply codegen/cmd/injection-gen/generators/packages.go:211]: This suggestion does not compile. vals["nonNamespaced"] is not boolean.

File: codegen/cmd/injection-gen/generators/comment_parser.go [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+			if opts == nil {
+				opts = make(map[string]string, len(vals))
+			}
+
+			for _, pair := range vals {
+				if kv := strings.SplitN(pair, "=", 2); len(kv) == 2 {
+					opts[kv[0]] = kv[1]
+				} else if len(kv) == 1 {
+					opts[kv[0]] = ""
+				}
```
Comment (vagababov) [inline codegen/cmd/injection-gen/generators/comment_parser.go:61]: Perhaps
```
if pair == "" {
  continue
}
```
then you don't need to check for `len(kv)==1` (and avoid calling split at all).

File: codegen/cmd/injection-gen/generators/comment_parser_test.go [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+func TestParseComments(t *testing.T) {
+	comment := []string{
+		"This is an example comment to parse",
+		"",
+		" +foo",
+		"+bar",
+		"+with:option",
+		"+pair:key=value",
+		"+manypairs:key1=value1,key2=value2",
+	}
```
Comment (vagababov) [inline codegen/cmd/injection-gen/generators/comment_parser_test.go:29]: so you can add a test, where: "a=b,,,,,c=d"

File: codegen/cmd/injection-gen/generators/packages.go (hunk 4 above)
Comment (vagababov) [reply codegen/cmd/injection-gen/generators/packages.go:211]: hm, for reason I thought it's a map to `bool`...

File: codegen/cmd/injection-gen/generators/comment_parser_test.go (hunk 11 above)
Comment (whaught) [reply codegen/cmd/injection-gen/generators/comment_parser_test.go:29]: Done

File: codegen/cmd/injection-gen/generators/comment_parser.go (hunk 10 above)
Comment (whaught) [reply codegen/cmd/injection-gen/generators/comment_parser.go]: I think the split is reasonable to detect the '=' char. We might be over-optimizing a bit a this point. This really just needs to work for happy cases - we can assume developers with malformed comments won't get generated code correctly.

**Comment by antoineco** [comment]
Prow says
> Code coverage is below threshold (50%), fail presubmit workflow intentionally

Probably due to the 0% coverage in `packages.go`, which is surprising because the file already existed in the repo before this PR, so coverage should be tackled separately.

File: codegen/cmd/injection-gen/generators/packages.go [hunk 12]
```diff
@@ -188,22 +188,27 @@ func MustParseClientGenTags(lines []string) Tags {
...
+	return ExtractCommentTags("+", comments)
+}
+
+func extractReconcilerClassTag(tags map[string]map[string]string) (classname string, has bool) {
+	vals, has := tags["genreconciler"]
+	if !has {
+		return
 	}
-	return "", false
+	classname, _ = vals["class"]
```
Comment (mattmoor) [inline codegen/cmd/injection-gen/generators/packages.go:201]: This doesn't properly set `has` anymore, and breaks serving.

Comment (mattmoor) [reply codegen/cmd/injection-gen/generators/packages.go:201]: I'm tweaking this, and will attempt to roll things forward

**Pull request merged**